keps/sig-network/4858-ip-cidr-validation/README.md

```diff
@@ -0,0 +1,838 @@
+
+  - Set by kubelet:
+    - `pod.status.hostIP`
+    - `pod.status.hostIPs[].ip`
+    - `pod.status.podIP`
+    - `pod.status.podIPs[].ip`
+
+  - Set by kube-apiserver:
+    - `service.spec.clusterIP`
+    - `service.spec.clusterIPs[]`
+
+  - Set by the cloud-provider framework code:
+    - `service.status.loadBalancer.ingress[].ip`
+
+We will make sure that the code paths that set these values always set
+them to fully-canonical values, even if they initially receive them in
+non-canonical form from some other component.
+
+However, we will still need stricter validation for these fields as
+well, since all of them are either mutable, or else sometimes set by
+external components.
+
+### API Warnings
+
+For several releases before changing the validation rules, the
+apiserver will return warnings when anyone sets any of the above
+fields to an invalid or non-canonical IP value:
+
+```
+spec.dnsConfig.nameservers[1]: non-standard IP address "05.06.07.08" will be considered invalid in a future Kubernetes release: use "5.6.7.8"
+
+spec.hostAliases[0].ip: non-standard IP address "::ffff:1.2.3.4" will be considered invalid in a future Kubernetes release: use "1.2.3.4"
+
+spec.loadBalancerSourceRanges[0]: CIDR value "192.12.2.8/24" is ambiguous in this context (should be "192.12.2.0/24" or "192.12.2.8/32"?)
+
+spec.clusterIPs[1]: IPv6 address "2001:db8:0:0::2" should be in RFC 5952 canonical format ("2001:db8::2")
+```
+
+### Updated Validation
+
+After several releases with the warnings, we will update the
+validation code to reject new invalid values. (We do not currently
+plan to reject non-canonical IPv6 addresses; that will remain a
+warning only.)
+
+All in-tree API types have already been modified to use appropriate
+`utilvalidation` methods for IP/CIDR validation ([kubernetes #122931],
+[kubernetes #123174]), so we just need to update the validation
+functions themselves to allow only unambiguous values.
+
+(UPDATE: the newly-added DRA `NetworkDeviceData` field uses
+`netutils.ParseCIDRSloppy` directly and needs to be fixed.)
+
+[kubernetes #122931]: https://github.com/kubernetes/kubernetes/pull/122931
+[kubernetes #123174]: https://github.com/kubernetes/kubernetes/pull/123174
+
+#### Ratcheting validation for pre-existing objects
+
+We want to allow making changes to objects containing invalid
+IPs/CIDRs, as long as the IPs/CIDRs themselves are not changed.
+
+In the case of the four fields above which are not arrays themselves,
+and not contained within arrays, this is simple: if `pod.spec.hostIP`,
+`pod.spec.podIP`, `node.status.podCIDR`, or `service.spec.clusterIP`
+contains an invalid value, and an update the the Pod/Service does not
+change the value of that field, then the invalid value should be
+allowed to remain.
+
+In the case of array-valued fields, this becomes more complicated; if
+a service has `clusterIPs: ["001.002.003.004"]`, should it be possible
+to update it to `clusterIPs: ["001.002.003.004", "fd00::1234"]`? Does
```

Comment:
What will happen in case of `clusterIPs: ["fd00::1234", "001.002.003.004"]`, which theoretically means I've removed the first element, and added a second which happens to be identical to the first old, but is currently second and invalid. Does it mean that old in this case also means exact position in the array?

Reply:
answered in the next paragraph: "In these cases, any invalid IP that appears anywhere in the array in the old object is allowed to appear anywhere in the array in the updated object."